server/lomas_server/utils/collection_models.py

```diff
@@ -88,11 +89,156 @@ class DatasetsCollection(BaseModel):
+
+
+class StrMetadata(ColumnMetadata):
+    """Model for string metadata"""
+
+    type: Literal["string"]
+
+
+class CategoricalColumnMetadata(ColumnMetadata):
+    """Model for categorical column metadata"""
+
+    @model_validator(mode="after")
+    def validate_categories(self):
+        """Makes sure number of categories matches cardinality."""
+        if len(self.categories) != self.cardinality:
+            raise ValueError(
+                "Number of categories should be equal to cardinality."
+            )
+        return self
+
+
+class StrCategoricalMetadata(CategoricalColumnMetadata):
+    """Model for categorical string metadata"""
+
+    type: Literal["string"]
+    cardinality: int
+    categories: List[str]
+
+
+class Precision(IntEnum):
+    """Precision of integer and float data"""
+
+    SINGLE = 32
+    DOUBLE = 64
+
+
+class BoundedColumnMetadata(ColumnMetadata):
+    """Model for columns with bounded data"""
+
+    @model_validator(mode="after")
+    def validate_bounds(self):
+        """Validates column bounds."""
+        if (
+            self.lower is not None
+            and self.upper is not None
+            and self.lower > self.upper
+        ):
+            raise ValueError("Lower bound cannot be larger than upper bound.")
+
+        return self
+
+
+class IntMetadata(BoundedColumnMetadata):
+    """Model for integer column metadata"""
+
+    type: Literal["int"]
+    precision: Precision
+    lower: int
+    upper: int
+
+
+class IntCategoricalMetadata(CategoricalColumnMetadata):
+    """Model for integer categorical column metadata"""
+
+    type: Literal["int"]
+    precision: Precision
+    cardinality: int
+    categories: List[int]
+
+
+class FloatMetadata(BoundedColumnMetadata):
+    """Model for float column metadata"""
+
+    type: Literal["float"]
+    precision: Precision
+    lower: float
+    upper: float
+
+
+class BooleanMetadata(ColumnMetadata):
+    """Model for boolean column metadata"""
+
+    type: Literal["boolean"]
+
+
+class DatetimeMetadata(BoundedColumnMetadata):
+    """Model for datetime column metadata"""
+
+    type: Literal["datetime"]  # TODO make these constants, see issue #268
+    lower: datetime
+    upper: datetime
+
+
+def get_column_metadata_discriminator(v: Any) -> str:
+    """Discriminator function for determining the type of column metadata.
+
+    Args:
+        v (Any): The unparsed column metadata (either dict or class object)
+
+    Raises:
+        ValueError: If the column type cannot be found.
+
+    Returns:
+        str: The metadata string type.
+    """
+    if isinstance(v, dict):
+        col_type = v.get("type")
+    else:
+        col_type = getattr(v, "type")
+
+    if (col_type in ("string", "int")) and (
+        ((isinstance(v, dict)) and "cardinality" in v)
+        or (hasattr(v, "cardinality"))
+    ):
+        col_type = f"categorical_{col_type}"
+
+    if not isinstance(col_type, str):
+        raise ValueError("Could not find column type.")
```

Comment:
InternalServerError ?

Reply:
Field validators in pydantic should use ValueError (or PydanticCustomError, which just changes the error type) and thus cannot use custom Exceptions.

Reply:
ok, but then what will happen in the server if it is raised ?

Reply:
Depends where. I think errors are catched in the server no? ie. it would raise an InternalServerException in the end.

I also parse the metadata dict into a Metadata object when adding it with mongo_admin.py for added safety.

Reply:
ok

Reply:
seems good